### PR TITLE
Enforce exclusion of hudson.* jenkins.* imports with some exemptions

### PR DESCRIPTION
**Pull request by i386**
# Description

Enforce exclusion of `hudson.*` `jenkins.*` imports with some exemptions

# Submitter checklist
- [ ] Link to JIRA ticket in description, if appropriate.
- [ ] Change is code complete and matches issue description
- [ ] Appropriate unit or acceptance tests or explanation to why this change has no tests
- [ ] Reviewer's manual test instructions provided in PR description. See Reviewer's first task below.
- [ ] Ran Acceptance Test Harness against PR changes.

# Reviewer checklist
- [ ] Run the changes and verified the change matches the issue description
- [ ] Reviewed the code
- [ ] Verified that the appropriate tests have been written or valid explanation given

### CODE REVIEW

File: blueocean-rest/src/main/java/io/jenkins/blueocean/Routable.java [hunk 1]
```diff
@@ -12,7 +10,7 @@
  */
 public interface Routable {
     /**
-     * Follows the same convention as {@link Action#getUrlName()}
+     * Follows the same convention as Action.getUrlName()
```
Comment (i386) [inline blueocean-rest/src/main/java/io/jenkins/blueocean/Routable.java]: Did not want to add `Action` to whitelist as nothing depends on it except for javadoc.

File: blueocean-rest/pom.xml [hunk 2]
```diff
@@ -24,4 +24,47 @@
...
+                                        <bannedImport>jenkins.**</bannedImport>
+                                        <bannedImport>hudson.**</bannedImport>
+                                    </bannedImports>
+                                    <allowedImports>
+                                        <allowedImport>hudson.Extension</allowedImport>
+                                        <allowedImport>hudson.ExtensionPoint</allowedImport>
+                                        <allowedImport>hudson.ExtensionList</allowedImport>
+                                        <allowedImport>hudson.ExtensionListListener</allowedImport>
+                                        <allowedImport>hudson.security.csrf.CrumbExclusion</allowedImport>
+                                        <allowedImport>hudson.util.AdaptedIterator</allowedImport>
```
Comment (i386) [inline blueocean-rest/pom.xml]: Needed for `Container`

File: blueocean-rest/pom.xml [hunk 3]
```diff
@@ -24,4 +24,47 @@
...
+                    <execution>
+                        <id>check-no-jenkins-imports</id>
+                        <phase>process-sources</phase>
+                        <goals>
+                            <goal>enforce</goal>
+                        </goals>
+                        <configuration>
+                            <rules>
+                                <restrictImports implementation="de.skuzzle.enforcer.restrictimports.RestrictImports">
+                                    <basePackage>**</basePackage>
```
Comment (i386) [inline blueocean-rest/pom.xml]: We could add a second rule to allow certain imports for a new `io.jenkins.blueocean.rest.factory` package. All factory classes would be moved from `blueocean-rest-impl` to `blueocean-rest`. Then plugins could safely depend on `blueocean-rest` and use the factories.

File: blueocean-rest/src/main/java/io/jenkins/blueocean/Routable.java (hunk 1 above)
Comment (vivek) [inline blueocean-rest/src/main/java/io/jenkins/blueocean/Routable.java]: +1

File: blueocean-rest/pom.xml (hunk 2 above)
Comment (vivek) [inline blueocean-rest/pom.xml]: Looks good. Wish these could be avoided.

File: blueocean-rest/pom.xml (hunk 3 above)
Comment (vivek) [inline blueocean-rest/pom.xml]: I guess we can do that, Item is one of the thing most blue extensions depend on, although it pulls in lots of other dependencies but not too bad. Where as Run (needed by BlueRunFactory) has too many dependencies and its like opening up to broader set of dependencies in Jenkins core.

**Comment by vivek** [comment]
LGTM 🐝 as long as test/ATH pass:)

**Pull request merged**